### DISCUSSION TITLE
Using double precision and origin offset in vehicle simulation

### DISCUSSION THREAD

**Discussion by fr0zi** [jrouwe/JoltPhysics#939] [Q&A] [answered]
Hello. I want to create some hobby project about vehicle simulation on a bigger terrain, so I would like to utilize double precision floats.
I read here on Discussions that there should be some methods that now takes "offset" as a parameter to shift the calculations to the origin.
Going through demos that comes with Jolt, I can see that for example some character classes have SetRootOffset method but I don't see anything like that for vehicles demo.

How to use this origin shift with vehicles? Thanks!

**Comment by jrouwe** [comment]
You don't need to shift the origin (ever). The vehicle positions are doubles, so you just fill in a position and let it run. You only need the 'base offset' in a few places, mostly collision queries. E.g. for NarrowPhaseQuery::CollideShape you provide it. All collision positions it returns (that are in floats) are relative to this 'base offset' so that you don't lose precision. But there is no such thing as THE origin, you can pick a different origin for every query, it's just a tool to get the most accurate results.

**Reply by fr0zi** [reply]
Yes, maybe my semantics were not 100% correct :) But as I understand the goal of this mechanism is the same as "origin shifting" found in other engines/libraries.
So vehicle's position is a double by default and I don't need any additional work here? That's cool.

**Reply by jrouwe** [reply, marked as answer]
> But as I understand the goal of this mechanism is the same as "origin shifting" found in other engines/libraries.

Not quite. What I would call 'origin shifting' (which I've seen in other engines) means that you explicitly tell the simulation around which origin the simulation should be located. If you want to shift that origin it internally effectively shifts all the bodies in the opposite direction (which takes a lot of CPU time). This is not how Jolt works, Jolt just stores positions in doubles.

**Reply by fr0zi** [reply]
I see. Ok, thanks for explanation. I thought internally it works the same or very similar to other engines. Now I understand the difference.
